src/main/java/org/apache/datasketches/filters/bloomfilter/BloomFilter.java

```diff
@@ -0,0 +1,747 @@
+          return false;
+    }
+    return true;
+  }
+
+  /**
+   * Returns the length of this BloomFilter when serialized, in bytes
+   * @return The length of this BloomFilter when serialized, in bytes
+   */
+  public long getSerializedSizeBytes() {
+    long sizeBytes = 2L * Long.BYTES; // basic sketch info + baseSeed
+    sizeBytes += bitArray_.getSerializedSizeBytes();
+    return sizeBytes;
+  }
+
+/*
+ * A Bloom Filter's serialized image always uses 3 longs of preamble, whether empty or not:
+ *
+ * <pre>
+ * Long || Start Byte Adr:
+ * Adr:
+ *      ||       0        |    1   |    2   |    3   |    4   |    5   |    6   |    7   |
+ *  0   || Preamble_Longs | SerVer | FamID  |  Flags |----Num Hashes---|-----Unused------|
+ *
+ *      ||       8        |    9   |   10   |   11   |   12   |   13   |   14   |   15   |
+ *  1   ||---------------------------------Hash Seed-------------------------------------|
+ *
+ *      ||      16        |   17   |   18   |   19   |   20   |   21   |   22   |   23   |
+ *  2   ||-------BitArray Length (in longs)----------|-----------Unused------------------|
+ *  </pre>
+ * 
+ * The raw BitArray bits, if non-empty start at byte 24.
+ */
+
+  /**
+   * Serializes the current BloomFilter to an array of bytes.
+   *
+   * <p>Note: Method throws if the serialized size exceeds <code>Integer.MAX_VALUE</code>.</p>
+   * @return A serialized image of the current BloomFilter as byte[]
+   */
+  public byte[] toByteArray() {
+    final long sizeBytes = getSerializedSizeBytes();
+    if (sizeBytes > Integer.MAX_VALUE) {
+      throw new SketchesStateException("Cannot serialize a BloomFilter of this size using toByteArray(); use toLongArray() instead.");
+    }
+
+    final byte[] bytes = new byte[(int) sizeBytes];
+    final WritableMemory wmem = WritableMemory.writableWrap(bytes);
+
+    int offsetBytes = 0;
+    wmem.putByte(offsetBytes++, (byte) Family.BLOOMFILTER.getMinPreLongs());
+    wmem.putByte(offsetBytes++, (byte) SER_VER); // to do: add constant
+    wmem.putByte(offsetBytes++, (byte) Family.BLOOMFILTER.getID());
+    wmem.putByte(offsetBytes++, (byte) (bitArray_.isEmpty() ? EMPTY_FLAG_MASK : 0));
+    wmem.putShort(offsetBytes, numHashes_);
+    offsetBytes += Integer.BYTES; // wrote a short and skipping the next 2 bytes
+    wmem.putLong(offsetBytes, seed_);
+    offsetBytes += Long.BYTES;
+
+    bitArray_.writeToMemory(wmem.writableRegion(offsetBytes, sizeBytes - offsetBytes));
+
+    return bytes;
+  }
+
+  /**
+   * Serializes the current BloomFilter to an array of longs. Unlike {@link #toByteArray()},
+   * this method can handle any size filter.
+   *
+   * @return A serialized image of the current BloomFilter as long[]
+   */
+  public long[] toLongArray() {
+    final long sizeBytes = getSerializedSizeBytes();
+
+    final long[] longs = new long[(int) (sizeBytes >> 3)];
+    final WritableMemory wmem = WritableMemory.writableWrap(longs);
```

Comment:
there is Buffer (mem.asBuffer()) that keeps track of the offset, so we don't have to